notebooks/03-query/11-joins.ipynb

```diff
@@ -0,0 +1,2475 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "93cb2f03-0831-41ae-8df8-865a77280aa2",
+   "metadata": {},
+   "source": [
+    "Run it 20 times..."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "id": "7b97cdfe-f46b-4cd9-8b59-faba4a4c015e",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Results = avg:1182.9867 ms   min:1093.643 ms  max:1455.442 ms\n",
+      "2023-10-26 01:00:29.929201 <<<<<<<<<<<<< \n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "datetime.datetime(2023, 10, 26, 1, 0, 29, 929201)"
+      ]
+     },
+     "execution_count": 14,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "join_results=[]\n",
+    "result = measure_query(sql, 20)\n",
+    "# saving all results for comparison \n",
+    "join_results.append({\"query_type\":\"simple_join\",\"results\":result})\n",
+    "print(result)\n",
+    "print_timestamp()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a040384d-3d46-4bce-a084-c020433fa676",
+   "metadata": {},
+   "source": [
+    "### Order of the Joins is Important\n",
+    "\n",
+    "![](assets/fail-scatter-gather.png)\n",
+    "\n",
+    "If you change the order of the joins, it will fail because the subquery will return more than `maxSubQueryRows` which defaults to 100000. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 15,
+   "id": "d8184fcf-a2cb-44b6-af0d-6bd9511fbd11",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "ERROR: Subquery generated results beyond maximum[100000] rows\n"
+     ]
+    }
+   ],
+   "source": [
+    "sql='''\n",
+    "SELECT TIME_FLOOR(c.__time, 'P1D') as _date, \n",
+    "  round( TIMESTAMPDIFF(YEAR, TIME_PARSE(u.dob), CURRENT_TIMESTAMP),-1) as age_group, \n",
+    "  count(distinct c.client_ip) as distint_ips\n",
+    "FROM users u\n",
+    "     INNER JOIN clicks c\n",
+    "        ON c.user_id=u.user_id\n",
+    "WHERE c.__time > CURRENT_TIMESTAMP  - INTERVAL '1' DAY\n",
+    "GROUP BY 1, round( TIMESTAMPDIFF(YEAR, TIME_PARSE(u.dob), CURRENT_TIMESTAMP),-1) \n",
+    "ORDER BY 3 DESC\n",
+    "LIMIT 10\n",
+    "'''\n",
+    "try:\n",
+    "  display.sql(sql)\n",
+    "except Exception as ex:\n",
+    "  print(f\"ERROR: {ex}\")"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "id": "92cb0f7e-60a1-4a7c-9060-08748cf35036",
+   "metadata": {},
+   "source": [
+    "### Nested Subqueries\n",
+    "See the query in the next cell, Druid will execute the subquery first as set of scatter/gather steps and then process the rest of the joins in subsequent scatter/gather steps. The results of the subquery still need to be within the `maxSubQueryRows` limit, but running it this way will allow all processing of the larger `clicks` data with the parallelism available on the Data Servers: \n",
```

Comment:
I had trouble unpicking the query here.

Maybe you could add some ```sql``` bits to explain what it's done?

You could even add multiple selects to show that the rowcount is lower than the config param.

Reply:
I added an explanation of what the query is doing and how the subquery is useful. We've already touched on the `maxSubqueryRows` aspect in a cell above and we mention that the subquery needs to be within the limit, do you still think we need to spell it out?